modules/stochastic_tools/src/utils/GaussianProcessHandler.C

```diff
@@ -245,6 +261,125 @@ GaussianProcessHandler::formFunctionGradient(Tao /*tao*/,
   *f = log_likelihood;
 }
 
+void
+GaussianProcessHandler::tuneHyperParamsADAM(RealEigenMatrix training_params,
+                                            RealEigenMatrix training_data,
+                                            const unsigned int & iter,
+                                            const unsigned int & batch_size,
+                                            const Real & learningRate)
+{
+  libMesh::PetscVector<Number> theta(_tao_comm, _num_tunable);
+  _batch_size = batch_size;
+  _covariance_function->buildHyperParamMap(_hyperparam_map, _hyperparam_vec_map);
+  mapToPetscVec(_tuning_data, _hyperparam_map, _hyperparam_vec_map, theta);
+  Real b1;
+  Real b2;
+  Real eps;
+  // Internal params for ADAM; set to the recommended values in the paper
+  b1 = 0.9;
+  b2 = 0.999;
+  eps = 1e-7;
+  std::vector<Real> m0;
+  std::vector<Real> v0;
+  for (unsigned int ii = 0; ii < _num_tunable; ++ii)
+  {
+    m0.push_back(0.0);
+    v0.push_back(0.0);
+  }
+  Real new_val;
+  Real m_hat;
+  Real v_hat;
+  Real store_loss;
+  std::vector<Real> grad1;
+
+  // Initialize randomizer
+  std::vector<unsigned int> v_sequence(training_params.rows());
+  std::iota(std::begin(v_sequence), std::end(v_sequence), 0);
+  RealEigenMatrix inputs(_batch_size, training_params.cols());
+  RealEigenMatrix outputs(_batch_size, 1);
+  Moose::out << "OPTIMIZING GP HYPER-PARAMETERS USING ADAM" << std::endl;
+  for (unsigned int ss = 0; ss < iter; ++ss)
+  {
+
+    // Shuffle data
+    MooseRandom generator;
+    generator.seed(0, 1980);
+    generator.saveState();
+    MooseUtils::shuffle<unsigned int>(v_sequence, generator, 0);
+    for (unsigned int ii = 0; ii < _batch_size; ++ii)
+    {
+      for (unsigned int jj = 0; jj < training_params.cols(); ++jj)
+        inputs(ii, jj) = training_params(v_sequence[ii], jj);
+      outputs(ii, 0) = training_data(v_sequence[ii], 0);
+    }
+    _training_params = &inputs;
```

Comment:
I suppose this is why you needed the change from const pointers to pointers in the header file. Why don't you just pass input+output to `getLossAdam` + `getGradientAdam` that way you don't have to copy the original database over and over again.